sdk/core/core/src/az_curl_adapter.c

```diff
@@ -10,27 +10,218 @@
+ * operations fails.
+ *
+ * @param writable_buffer
+ * @param url_from_request
+ * @return az_result
+ */
+az_result az_write_url(az_span const writable_buffer, az_const_span const url_from_request) {
+  az_span_builder writer = az_span_builder_create(writable_buffer);
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, url_from_request));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(&writer, AZ_STR("\0")));
+  return AZ_OK;
+}
+
+/**
+ * handles GET request
+ */
+az_result az_curl_send_request(az_curl * const p_curl, az_http_request_builder * const p_hrb) {
+  // creates a slist for bulding curl headers
+  az_curl_headers_list headers = {
+    .p_list = NULL,
+  };
+  // build headers into a slist as curl is expecting
+  AZ_RETURN_IF_FAILED(az_build_headers(p_hrb, &headers));
+  // set all headers from slist
+  curl_easy_setopt(p_curl->p_curl, CURLOPT_HTTPHEADER, headers.p_list);
 
-az_result az_headers_to_curl(az_headers_data * const p_state, az_pair const header) {
-  const az_span_seq token_seq = az_build_header_callback(&header);
-  char * str_header;
-  AZ_RETURN_IF_FAILED(az_span_seq_to_new_str(token_seq, &str_header));
+  // set URL
+  uint8_t * const p_writable_buffer = (uint8_t * const)malloc(p_hrb->url.size + 1);
+  char * buffer = (char *)p_writable_buffer;
+  az_span const writable_buffer
+      = (az_span const){ .begin = p_writable_buffer, .size = p_hrb->url.size + 1 };
+  az_result result = az_write_url(writable_buffer, az_span_to_const_span(p_hrb->url));
+  if (az_succeeded(result)) {
+    curl_easy_setopt(p_curl->p_curl, CURLOPT_URL, buffer);
+  }
+  memset(p_writable_buffer, 0, p_hrb->url.size + 1);
+  free(buffer);
 
-  p_state->p_list = curl_slist_append(p_state->p_list, str_header);
+  // send
+  CURLcode res = curl_easy_perform(p_curl->p_curl);
+  if (res != CURLE_OK) {
+    fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
+  }
+
+  // TODO: would this be part of done function instead?
+  curl_slist_free_all(headers.p_list);
 
-  free(str_header);
   return AZ_OK;
 }
 
-az_result az_build_headers(az_http_request const * const p_request, az_headers_data * p_headers) {
-  // create callback for visitor
-  az_pair_visitor const pair_visitor = az_headers_to_curl_callback(p_headers);
+/**
+ * @brief This is the function that curl will use to write response into a user provider span
+ * Function receives the size of the response and must return this same number, otherwise it is
+ * consider that function failed
+ *
+ * @param contents
+ * @param size
+ * @param nmemb
+ * @param userp
+ * @return int
+ */
+int write_to_span(void * contents, size_t size, size_t nmemb, void * userp) {
+  size_t realsize = size * nmemb + 1;
+  az_span * const mem = (az_span * const)userp;
+
+  if (mem->size == 0) {
+    mem->begin = malloc(realsize);
+    mem->size = realsize;
+  } else if (mem->size < realsize) {
+    uint8_t * ptr = realloc(mem->begin, realsize);
```

Comment:
For response we should either use a buffer or use a callback. But we shouldn't reallocate, even for CURL.

Reply:
why?
If we are already allocating for request, why the response should've been different?

Reply:
Because we can't return to a user dynamically allocated buffer.